File: server/src/main/java/net/progruzovik/dissent/captain/AbstractCaptain.java
package net.progruzovik.dissent.captain;

import net.progruzovik.dissent.battle.model.Battle;
import net.progruzovik.dissent.battle.model.Side;
import net.progruzovik.dissent.model.entity.Ship;
import org.springframework.lang.NonNull;
import org.springframework.lang.Nullable;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractCaptain implements Captain {

    private final List<Ship> ships = new ArrayList<>();
    private Battle battle;

    @NonNull
    @Override
    public List<Ship> getShips() {
        return ships;
    }

    @Nullable
    @Override
    public Battle getBattle() {
        return battle;
    }

    @Override
    public void addToBattle(@NonNull Side side, @NonNull Battle battle) {
        battle.addObserver(this);
        battle.addShips(side, ships);
        this.battle = battle;
    }
}

File: server/src/main/kotlin/net/progruzovik/dissent/captain/AbstractCaptain.kt
package net.progruzovik.dissent.captain

import net.progruzovik.dissent.model.domain.Ship
import net.progruzovik.dissent.model.domain.battle.Battle
import net.progruzovik.dissent.model.domain.battle.Side

abstract class AbstractCaptain : Captain {

    override val ships = ArrayList<Ship>()
    override var battle: Battle? = null

    override fun addToBattle(battle: Battle, side: Side) {
        battle.addShips(side, ships)
        this.battle = battle
    }

    override fun endTurn() {
        battle?.endTurn(id)
    }
}
